Sources/SourceKitLSP/DocumentManager.swift

```diff
@@ -98,43 +116,99 @@ public final class DocumentManager {
 
   /// Applies the given edits to the document.
   ///
-  /// - parameter editCallback: Optional closure to call for each edit.
+  /// - parameter willEditDocument: Optional closure to call before each edit.
+  /// - parameter updateDocumentTokens: Optional closure to call after each edit.
   /// - parameter before: The document contents *before* the edit is applied.
+  /// - parameter after: The document contents *after* the edit is applied.
   /// - returns: The contents of the file after all the edits are applied.
   /// - throws: Error.missingDocument if the document is not open.
   @discardableResult
-  public func edit(_ uri: DocumentURI, newVersion: Int, edits: [TextDocumentContentChangeEvent], editCallback: ((_ before: DocumentSnapshot, TextDocumentContentChangeEvent) -> Void)? = nil) throws -> DocumentSnapshot {
+  public func edit(
+    _ uri: DocumentURI,
+    newVersion: Int,
+    edits: [TextDocumentContentChangeEvent],
+    willEditDocument: ((_ before: DocumentSnapshot, TextDocumentContentChangeEvent) -> Void)? = nil,
+    updateDocumentTokens: ((_ after: DocumentSnapshot) -> DocumentTokens?)? = nil
```

Comment:
I think `updateDocumentTokens` can return a non-optional now (see https://github.com/apple/sourcekit-lsp/pull/414#discussion_r687469166)
```suggestion
    updateDocumentTokens: ((_ after: DocumentSnapshot) -> DocumentTokens)? = nil
```